icu4c/source/data/brkitr/ja.txt

```diff
@@ -13,13 +13,47 @@ ja{
         line_phrase:process(dependency){"line_phrase_cj.brk"}
     }
     particles{
-         か, かしら, から, が, くらい, けれども, こそ,
-         さ, さえ, しか, だけ, だに, だの, て, で, でも,
-         と, ところが, とも, な, など, なり, に, ね, の,
-         ので, のに, は, ば, ばかり, へ, ほど, まで, も,
-         や, やら, よ, より, わ, を
-    }
-    auxVerbs {
-        です, でしょ, でし, ます, ませ, まし
+        々宮,々家,えたい,えて,える,
```

Comment:
How is this list generated?

Reply:
Hiro helped extract from cjdict.txt

Reply:
I am very worry about this adding more data entries approach. Are these entries based on some standard? some text book? some reference manual? How can we justify this is a good list but not a bad list? If we find a highschool Japanese teacher in Japan and ask him/her to provide the list, would he/she give us a different list? I do not feel there are good refererence material to prove this list is good (or bad) @markusicu @pedberg-icu

Reply:
Hi Frank,

As Allen mentioned, all entries come from cjdict.txt. So they are based on the current ICU implementation.

Some of the entries are actual particles which are in text books (e.g. かしら, から)
Some of the entries are not particles or even words, but they are in cjdict.txt (e.g. 々宮, ーン)

We need to add those entries, even if they are not proper words, unless they are in cjdict.txt.
This is because the current word wrapping implementation breaks sentences with cjdict.txt to words first, then merge the words with entries in the particles.

Ideally we should remove those entries from both cjdict.txt and particles, and also add proper words to the both lists. However, it may affect some compatibility issues and it requires a long time to address them.

So how about adding comments as follows?
* All entries are extracted from cjdict.txt.
* When cjdict.txt is modified, this list should be updated too.

Thanks,

Reply:
It sounds like the extra entries are for "words" which we have in the dictionary but shouldn't have; and that we need to add them here so that they can be re-attached to the preceding Kanji word after the regular dictionary code split them. Right? Assuming that...

1. Please make sure that at least a couple of words with these extra entries are covered by the test data.
2. Then, in order for us (as a team) being able to maintain this data going forward, we need to have permanent, findable documentation about which entries are "actual particles which are in text books" vs. which ones are added for technical reasons/limitations. The best way to do this might be a comment right here in the .txt file, and separating the two sub-lists: a comment for what's going on, then the "actual particles" (ideally with a reference to some standard or educational material etc.), then a separator comment, then the other entries.
3. Please submit an ICU ticket for reviewing these extra entries in the dictionary, and removing as many of them as possible.
4. And please remember this when you work on the follow-up task of generating this modified file.

Reply:
> 1. Please make sure that at least a couple of words with these extra entries are covered by the test data.

Allen, what about the following test cases?
```
しかし、これらのサービスは６年間使ってこなかった自分にとっては特に必要でないものであり、これからこの機能が加わったからといって特別ハッピーなわけでもない。
自由が丘での三ヶ月の生活を通して得られる経験
```

You can see the details in the following spreadsheet:
https://docs.google.com/spreadsheets/d/1MYvDEuXqDeDSUM-EmS3d4DZytMF3aBfu6YTuU_wzDkQ/edit#gid=1340515435

> 2. Then, in order for us (as a team) being able to maintain this data going forward...

I summarized the categories of all entries:
https://docs.google.com/spreadsheets/d/1MYvDEuXqDeDSUM-EmS3d4DZytMF3aBfu6YTuU_wzDkQ/edit#gid=0

* `particle` is particle which is in text books (e.g. から).
* `noun suffix` is a part of noun (e.g. ヶ丘) This is like "-tion". This is added for technical reasons.
* `particle+others` is a compound of particle and other words. (e.g. をもって) This is added for technical reasons.

Reply:
Thanks for adding some test data.

I would like @allensu05 and @hiroyuki-komatsu to agree on the list of strings here, and on the tests. Please reply on this pull request when you have come to an agreement.

I would like to see the comments to be a bit more self-explanatory, so that others reading & maintaining this data know where it comes from and have a chance to adjust it as appropriate. Including references to the ICU cjdict file if that's a reason why a group of strings has been added, and a reference for the "actual particles".

Copied from above so that it does not get lost: Please submit an ICU ticket for reviewing these extra entries in the dictionary, and removing as many of them as possible.

Then I can rubber-stamp and merge this pull request.

And then please work with @pedberg-icu on moving the data, *and the comments*, to CLDR so that the CLDR-to-ICU conversion generates an equivalent .txt file. I assume that at that point the comments will live only in the CLDR .xml file, and that the conversion drops them; if it's easy to preserve the comments, then that would be useful, but if it's hard then so be it.

Reply:
I am hoping this can get resolved and merged fairly soon so we can add the particles data in CLDR and use it during regular CDLR-ICU conversion to generate icu4c/source/data/brkitr/ja.txt. The CLDR data is on an earlier schedule than the ICU feature/code freezes so sooner would be better...thanks.

Reply:
> I would like @allensu05 and @hiroyuki-komatsu to agree on the list of strings here, and on the tests. Please reply on this pull request when you have come to an agreement.

@allensu05 and I (@hiroyuki-komatsu) agreed on the list of strings and the tests.

> I would like to see the comments to be a bit more self-explanatory, so that others reading & maintaining this data know where it comes from and have a chance to adjust it as appropriate. Including references to the ICU cjdict file if that's a reason why a group of strings has been added, and a reference for the "actual particles".

How about these comments?
```
// Particles
↓
// Entries in cjdict.txt categorized as particles (助詞), auxiliary verbs (助動詞),
// conjugations (活用)  and prefix of them.
// These entries are connected to the previous words for phrase based word segmentation.
// https://ja.wikipedia.org/wiki/助詞
// https://ja.wikipedia.org/wiki/助動詞_(国文法)
// https://ja.wikipedia.org/wiki/活用

// entries added for technical reasons (particle + others)
↓
// Entries in cjdict.txt, which are compound words starting from the above categories
// (i.e. particles, auxiliary verbs and conjunctions).
// These entries are connected to the previous words for phrase based word segmentation.

// entries added for technical reasons (noun suffix)
↓
// Entries in cjdict.txt, which are suffixes of nouns.
// These entries should be connected to the previous (noun) words.
```

Allen is updating the file.

> Copied from above so that it does not get lost: Please submit an ICU ticket for reviewing these extra entries in the dictionary, and removing as many of them as possible.

If my understanding is correct, we have a conclusion to put those entries into [icu4c/source/data/xml/brkitr/ja.xml](https://github.com/unicode-org/icu/blob/main/icu4c/source/data/xml/brkitr/ja.xml) in another thread. (Because ja.txt is generated from CLDR and ja.xml, and the data is tightly connected with cjdict.txt and not so sharable for other purposes).

So, as a next step, should Allen and I submit an ICU ticket for the modification of ja.xml as a separate change from this PR, or is it OK to keep this PR with the ja.xml change?


> And then please work with @pedberg-icu on moving the data, and the comments, to CLDR so that the CLDR-to-ICU conversion generates an equivalent .txt file. I assume that at that point the comments will live only in the CLDR .xml file, and that the conversion drops them; if it's easy to preserve the comments, then that would be useful, but if it's hard then so be it.

Allen is now working on ldmlICU.dtd and CLDR-ICU tools.

Thank you,

Reply:
> > I would like to see the comments to be a bit more self-explanatory, so that others reading & maintaining this data know where it comes from and have a chance to adjust it as appropriate. Including references to the ICU cjdict file if that's a reason why a group of strings has been added, and a reference for the "actual particles".
> 
> How about these comments? [...]

These comments lgtm, but I don't yet see them in the file.

> > Copied from above so that it does not get lost: Please submit an ICU ticket for reviewing these extra entries in the dictionary, and removing as many of them as possible.
> 
> If my understanding is correct, we have a conclusion to put those entries into [icu4c/source/data/xml/brkitr/ja.xml](https://github.com/unicode-org/icu/blob/main/icu4c/source/data/xml/brkitr/ja.xml) in another thread. (Because ja.txt is generated from CLDR and ja.xml, and the data is tightly connected with cjdict.txt and not so sharable for other purposes).
> 
> So, as a next step, should Allen and I submit an ICU ticket for the modification of ja.xml as a separate change from this PR, or is it OK to keep this PR with the ja.xml change?

Please do that as part of this ticket.

And please consider renaming the "particles" resource bundle key to something like "extensions" in this PR or in the move-data-to-xml PR, as Peter requested.

My request above was for something else. You mentioned before that some of the cjdict entries should not be in cjdict, and that you add "particles" here only because those strings are in cjdict. So please submit an ICU ticket describing the extraneous cjdict entries, for reviewing them, and for removing them from both cjdict and the "particles" list.

Reply:
Hi Markus
I've added comments into ja.txt. Please help review again.

In the mail "The new "particles" data for Japanese line break", two items to confirm:
1. Peter said he will help with the conversion from ja.xml to ja.txt "after this PR is merged". If so, I'll not do it in this PR.
2. Hiro said he preferred to keep the cjdict.txt content because one entry's removal may result in several entries' addition.
    Do you have concern about this?

Thanks

Reply:
I just submitted a new ICU ticket: https://unicode-org.atlassian.net/browse/ICU-21915
Please comment there if I misunderstood, or if you can add details.